bin/propolis-server/src/lib/vm/ensure.rs

```diff
@@ -236,39 +236,70 @@ impl<'a> VmEnsureNotStarted<'a> {
             },
         ));
 
-        // Create the runtime that will host tasks created by VMM components
-        // (e.g. block device runtime tasks).
-        let vmm_rt = tokio::runtime::Builder::new_multi_thread()
-            .thread_name("tokio-rt-vmm")
-            .worker_threads(usize::max(
-                VMM_MIN_RT_THREADS,
-                VMM_BASE_RT_THREADS + spec.board.cpus as usize,
-            ))
-            .enable_all()
-            .build()?;
-
         let log_for_init = self.log.clone();
         let properties = self.ensure_request.properties.clone();
         let options = self.ensure_options.clone();
         let queue_for_init = input_queue.clone();
-        let init_result = vmm_rt
-            .spawn(async move {
-                initialize_vm_objects(
-                    log_for_init,
-                    spec,
-                    properties,
-                    options,
-                    queue_for_init,
-                )
-                .await
-            })
-            .await
-            .map_err(|e| {
-                anyhow::anyhow!("failed to join VM object creation task: {e}")
-            })?;
-
-        match init_result {
-            Ok(objects) => {
+
+        // Either the block following this succeeds with both a Tokio runtime
+        // and VM objects, or entirely fails with no partial state for us to
+        // clean up.
+        type InitResult =
+            anyhow::Result<(tokio::runtime::Runtime, InputVmObjects)>;
+
+        // We need to create a new runtime to host the tasks for this VMM's
+        // objects, but that initialization is fallible and results in dropping
+        // the fledgling VMM runtime itself. Dropping a Tokio runtime on a
+        // worker thread in a Tokio runtime will panic, so do all init in a
+        // `spawn_blocking` where this won't be an issue.
+        //
+        // When the runtime is returned to this thread, it must not be dropped.
+        // That means that the path between this result and returning an
+        // `Ok(VmEnsureObjectsCreated)` must be infallible.
+        //
+        // `VmEnsureObjectsCreated` (and later state transitions) take care to
+        // `shutdown_background` the runtime.
```

Comment:
nit: this isn't quite right, I think; once you have a `VmEnsureActive` you are assured that the runtime will be shut down in the background when it's dropped. You get that by calling `VmEnsureObjectsCreated::ensure_active`.

We might want to put a `WARNING:` comment on `VmEnsureObjectsCreated` indicating that it can't currently be dropped (or maybe just add a Drop impl for it that calls `shutdown_background`, though I would have to convince myself that this won't leak any tasks).

Reply:
oh, i'd conflated `ActiveVm` and `VmEnsureActive`, yeah you're right.

i'll go with a stern warning. a Drop impl that calls `shutdown_background` _would_ immediately leak any object's tasks, but they're all also cancelled so they should stop "soon", so i think it's fine. but being in a situation where we're dropping `VmEnsureObjectsCreated` itself feels like a bug. _and_ if you innocently destructure `VmEnsureObjectsCreated` outside `ensure_active` you still have to be careful about the runtime again..

Reply:
(https://github.com/oxidecomputer/propolis/pull/866/commits/7053acf3b599981df0f471107bf18060145f329a)